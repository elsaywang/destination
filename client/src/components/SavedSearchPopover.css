:local(.heading) {
    margin-bottom: 0;
}

:local(.block) {
    margin: 0 0 10px;
}

#isCurrentSearch {
    background: var(--spectrum-global-color-blue500);
    color: var(--spectrum-global-color-grey50);
    border-color: var(--spectrum-global-color-blue500);
}
