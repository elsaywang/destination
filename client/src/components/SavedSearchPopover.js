import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { injectIntl, intlShape } from 'react-intl';
import { Tag } from '@react/react-spectrum/TagList';
import OverlayTrigger from '@react/react-spectrum/OverlayTrigger';
import ModalTrigger from '@react/react-spectrum/ModalTrigger';
import Dialog from '@react/react-spectrum/Dialog';
import Popover from '@react/react-spectrum/Popover';
import FieldLabel from '@react/react-spectrum/FieldLabel';
import { formatSignal } from '../utils/stringifySignals';
import { getDateRangeLabel } from '../constants/dateRangeOptions';
import { getSignalStatusLabel } from '../constants/signalStatusOptions';
import { getSignalTypeLabel } from '../constants/signalTypeOptions';
import { getSignalCategory } from '../constants/signalCategoryOptions';
import { getSortLabel } from '../constants/sortOptions';
import styles from './SavedSearchPopover.css';
import Heading from '@react/react-spectrum/Heading';
import Button from '@react/react-spectrum/Button';
import DeleteOutline from '@react/react-spectrum/Icon/DeleteOutline';

class SavedSearchPopover extends Component {
    onSavedSearchClick = search => {
        this.props.onSavedSearchClick(search);
    };

    getCustomDateRangeLabel() {
        const { search, intl } = this.props;
        const { customStartDate, customEndDate } = search;
        const formatDateOptions = {
            year: 'numeric',
            month: 'long',
            day: 'numeric',
            timeZone: 'UTC',
        };

        return `${intl.formatDate(customStartDate, formatDateOptions)} to ${intl.formatDate(
            customEndDate,
            formatDateOptions,
        )}`;
    }

    getViewRecordsForLabel() {
        const { search } = this.props;
        const isCustomDateRangeEnabled = search.viewRecordsFor === 'custom';

        return isCustomDateRangeEnabled
            ? this.getCustomDateRangeLabel()
            : getDateRangeLabel(search.viewRecordsFor);
    }

    getSignalSourceLabel(source) {
        if (source.reportSuiteIds.length) {
            return 'Report Suite';
        }

        if (source.dataSourceIds.length) {
            return 'Onboarded Record';
        }

        return '';
    }

    renderSignalSource(source) {
        return source.reportSuiteIds.length || source.dataSourceIds.length ? (
            <FieldLabel position="left" label={this.getSignalSourceLabel(source)}>
                <span style={{ verticalAlign: 'bottom' }}>{source.name}</span>
            </FieldLabel>
        ) : null;
    }

    render() {
        const { search, deleteSearch, isCurrentSearch } = this.props;
        const boundDeleteClick = deleteSearch.bind(this, search);

        return (
            <span className={styles.tagBlock}>
                <OverlayTrigger trigger="hover" data-test="saved-search-overlay-trigger">
                    <Tag
                        data-test="saved-search-tag"
                        id={isCurrentSearch ? 'isCurrentSearch' : null}
                        className={styles.tag}
                        key={search.name}
                        value={search.name}
                        onClick={() => this.onSavedSearchClick(search)}
                    />
                    <Popover title={search.name} data-test="saved-search-popover">
                        <Heading size="5" className={styles.heading}>
                            Name
                        </Heading>
                        <div>{search.name}</div>

                        <Heading size="5" className={styles.heading}>
                            Search Query
                        </Heading>
                        <div>{formatSignal(search)}</div>

                        <Heading size="5" className={styles.heading}>
                            Filters
                        </Heading>

                        <div>
                            <FieldLabel position="left" label="Signal Status">
                                <span style={{ verticalAlign: 'bottom' }}>
                                    {getSignalStatusLabel(search.signalStatus)}
                                </span>
                            </FieldLabel>

                            <FieldLabel position="left" label="Signal Category">
                                <span style={{ verticalAlign: 'bottom' }}>
                                    {getSignalCategory(search.source.sourceType)}
                                </span>
                            </FieldLabel>

                            {search.source.sourceType && (
                                <FieldLabel position="left" label="Signal Type">
                                    <span style={{ verticalAlign: 'bottom' }}>
                                        {getSignalTypeLabel(search.source.sourceType)}
                                    </span>
                                </FieldLabel>
                            )}

                            {search.source.sourceType && this.renderSignalSource(search.source)}

                            <FieldLabel position="left" label="View Records For">
                                <span style={{ verticalAlign: 'bottom' }}>
                                    {this.getViewRecordsForLabel()}
                                </span>
                            </FieldLabel>
                        </div>

                        <Heading size="5" className={styles.heading}>
                            Sorting
                        </Heading>
                        <div>{getSortLabel(search.sortBy)}</div>
                    </Popover>
                </OverlayTrigger>
                {isCurrentSearch && (
                    <ModalTrigger>
                        <Button
                            data-test="saved-search-delete-button"
                            className={styles.button}
                            label={null}
                            variant="action"
                            icon={<DeleteOutline size="XS" />}
                            modaltrigger
                        />
                        <Dialog
                            data-test="saved-search-delete-modal"
                            modalcontent
                            title="Delete Saved Search"
                            confirmLabel="Delete"
                            size="S"
                            cancelLabel="Cancel"
                            onConfirm={boundDeleteClick}
                            variant="destructive">
                            <span>
                                Are you sure you want to delete this saved search: {search.name}?
                            </span>
                        </Dialog>
                    </ModalTrigger>
                )}
            </span>
        );
    }
}

SavedSearchPopover.propTypes = {
    search: PropTypes.object.isRequired,
    onSavedSearchClick: PropTypes.func.isRequired,
    isCurrentSearch: PropTypes.bool.isRequired,
    deleteSearch: PropTypes.func.isRequired,
    intl: intlShape.isRequired,
};

export default injectIntl(SavedSearchPopover);
