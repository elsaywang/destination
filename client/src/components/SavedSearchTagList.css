:local(.noSavedSearches) {
    line-height: 30px;
    margin-right: 10px;
}
