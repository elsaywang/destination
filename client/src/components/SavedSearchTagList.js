import React from 'react';
import PropTypes from 'prop-types';
import { TagList, Tag } from '@react/react-spectrum/TagList';

function SavedSearchTagList({ list, onSavedSearchClick }) {
    const boundClick = search => onSavedSearchClick.bind(this, search);
    const renderTag = search => (
        <Tag
            data-test="saved-search-tag"
            style={{ cursor: 'pointer' }}
            key={search.name}
            value={search.name}
            onClick={boundClick}
        />
    );

    return (
        <TagList data-test="saved-search-tag-list" readOnly>
            {list.map(renderTag)}
        </TagList>
    );
}

SavedSearchTagList.propTypes = {
    list: PropTypes.array.isRequired,
    onSavedSearchClick: PropTypes.func.isRequired,
};

export default SavedSearchTagList;
