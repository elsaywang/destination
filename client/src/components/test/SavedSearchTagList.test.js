import React from 'react';
import { shallow } from 'enzyme';
import SavedSearchTagList from '../../components/SavedSearchTagList';
import { TagList } from '@react/react-spectrum/TagList';
import SavedSearchPopover from '../../components/SavedSearchPopover';
import Heading from '@react/react-spectrum/Heading';

describe('<SavedSearchTagList /> component', () => {
    const mockFn = jest.fn();
    const list = [
        {
            id: 17966,
            name: 'Germaine Wiza',
            keyValuePairs: [
                {
                    key: 'k-monitor',
                    operator: '>',
                    value: 'v-programming',
                },
            ],
            type: 'Lek',
            source: 'Buckinghamshire',
            totalCounts: 59925,
            percentageChange: 45886,
            includedTraits: [0, 1, 2],
            signalStatus: 'USED',
        },
    ];
    const wrapper = shallow(<SavedSearchTagList list={list} onSavedSearchClick={mockFn} />);

    describe('rendering', () => {
        it('matches snapshot', () => {
            expect(wrapper).toMatchSnapshot();
        });

        it('renders <TagList /> with a list of <Tag /> rendered according to the list passed in', () => {
            const tagList = wrapper.find(TagList);

            expect(tagList.filter('[data-test="saved-search-tag-list"]').exists()).toBe(true);
            expect(tagList.find(SavedSearchPopover).length).toBe(list.length);
        });
    });
});
