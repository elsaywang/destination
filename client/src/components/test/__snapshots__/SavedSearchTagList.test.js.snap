// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SavedSearchTagList /> component rendering matches snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SavedSearchTagList
    currentSearch="Watson Cartwright"
    list={
      Array [
        Object {
          "id": 17966,
          "includedTraits": Array [
            0,
            1,
            2,
          ],
          "keyValuePairs": Array [
            Object {
              "key": "k-monitor",
              "operator": ">",
              "value": "v-programming",
            },
          ],
          "name": "Germaine Wiza",
          "percentageChange": 45886,
          "signalStatus": "USED",
          "source": "Buckinghamshire",
          "totalCounts": 59925,
          "type": "Lek",
        },
      ]
    }
    onSavedSearchClick={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <SavedSearchPopover
          isCurrentSearch={false}
          onSavedSearchClick={[MockFunction]}
          search={
            Object {
              "id": 17966,
              "includedTraits": Array [
                0,
                1,
                2,
              ],
              "keyValuePairs": Array [
                Object {
                  "key": "k-monitor",
                  "operator": ">",
                  "value": "v-programming",
                },
              ],
              "name": "Germaine Wiza",
              "percentageChange": 45886,
              "signalStatus": "USED",
              "source": "Buckinghamshire",
              "totalCounts": 59925,
              "type": "Lek",
            }
          }
        />,
      ],
      "data-test": "saved-search-tag-list",
      "disabled": false,
      "invalid": false,
      "onBlur": [Function],
      "onClose": [Function],
      "onFocus": [Function],
      "readOnly": true,
      "required": false,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "Germaine Wiza",
        "nodeType": "function",
        "props": Object {
          "isCurrentSearch": false,
          "onSavedSearchClick": [MockFunction],
          "search": Object {
            "id": 17966,
            "includedTraits": Array [
              0,
              1,
              2,
            ],
            "keyValuePairs": Array [
              Object {
                "key": "k-monitor",
                "operator": ">",
                "value": "v-programming",
              },
            ],
            "name": "Germaine Wiza",
            "percentageChange": 45886,
            "signalStatus": "USED",
            "source": "Buckinghamshire",
            "totalCounts": 59925,
            "type": "Lek",
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <SavedSearchPopover
            isCurrentSearch={false}
            onSavedSearchClick={[MockFunction]}
            search={
              Object {
                "id": 17966,
                "includedTraits": Array [
                  0,
                  1,
                  2,
                ],
                "keyValuePairs": Array [
                  Object {
                    "key": "k-monitor",
                    "operator": ">",
                    "value": "v-programming",
                  },
                ],
                "name": "Germaine Wiza",
                "percentageChange": 45886,
                "signalStatus": "USED",
                "source": "Buckinghamshire",
                "totalCounts": 59925,
                "type": "Lek",
              }
            }
          />,
        ],
        "data-test": "saved-search-tag-list",
        "disabled": false,
        "invalid": false,
        "onBlur": [Function],
        "onClose": [Function],
        "onFocus": [Function],
        "readOnly": true,
        "required": false,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "Germaine Wiza",
          "nodeType": "function",
          "props": Object {
            "isCurrentSearch": false,
            "onSavedSearchClick": [MockFunction],
            "search": Object {
              "id": 17966,
              "includedTraits": Array [
                0,
                1,
                2,
              ],
              "keyValuePairs": Array [
                Object {
                  "key": "k-monitor",
                  "operator": ">",
                  "value": "v-programming",
                },
              ],
              "name": "Germaine Wiza",
              "percentageChange": 45886,
              "signalStatus": "USED",
              "source": "Buckinghamshire",
              "totalCounts": 59925,
              "type": "Lek",
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
